### PR TITLE
perf: independent parallel building for IVF_HNSW partitions

### PR DESCRIPTION
**Pull request by BubbleCal**
This improves the index building by 2x faster,
also fixes a panic bug caused by passing wrong offset to read batch.

### CODE REVIEW

File: rust/lance-index/src/vector/pq/storage.rs [hunk 1]
```diff
@@ -413,11 +413,12 @@ impl ProductQuantizationStorage {
         writer: &mut FileWriter<ManifestDescribing>,
     ) -> Result<usize> {
         let batch_size: usize = 10240; // TODO: make it configurable
-        for i in (0..self.batch.num_rows()).step_by(batch_size) {
-            let offset = i * batch_size;
+        let mut offset = 0;
+        while offset < self.batch.num_rows() {
```
Comment (eddyxu) [inline rust/lance-index/src/vector/pq/storage.rs:417]: is this the same as `for offset in (0..self.batch_num_rows().step_by(batch_size)` ,and just don't do `let offset = i * batch_size`?

File: rust/lance/src/index/vector/ivf/io.rs [hunk 2]
```diff
@@ -309,58 +320,190 @@ pub(super) async fn write_hnsw_index_partitions(
...
+            aux_part_writer,
+            pq,
+            row_id_array,
+            pq_array,
+        ));
     }
 
-    while let Some(task) = task_queue.pop_front() {
-        let (hnsw, pq_storage) = task.await?;
+    let parallel_limit = 16;
```
Comment (eddyxu) [inline rust/lance/src/index/vector/ivf/io.rs:363]: How do we pick 16 here? Btw lets define this as `const`?

File: rust/lance/src/index/vector/ivf/io.rs [hunk 3]
```diff
@@ -309,58 +320,190 @@ pub(super) async fn write_hnsw_index_partitions(
...
     metric_type: MetricType,
     hnsw_params: Arc<HnswBuildParams>,
+    mut writer: FileWriter<ManifestDescribing>,
+    mut aux_writer: Option<FileWriter<ManifestDescribing>>,
+    pq: Arc<dyn ProductQuantizer>,
+    row_id_array: Vec<Arc<dyn Array>>,
+    pq_array: Vec<Arc<dyn Array>>,
+) -> Result<usize> {
+    let projection = Arc::new(dataset.schema().project(&[column.as_ref()])?);
+    // let vector_batches = futures::stream::iter(row_id_array.iter())
```
Comment (eddyxu) [inline rust/lance/src/index/vector/ivf/io.rs:439]: Remove this code?

File: rust/lance/src/index/vector/ivf/io.rs [hunk 4]
```diff
@@ -309,58 +320,190 @@ pub(super) async fn write_hnsw_index_partitions(
         )
         .await?;
 
-        let mut vector_batches = Vec::new();
+        let writer_options = FileWriterOptions::default();
+        let (part_file, aux_part_file) = build_part_path(part_id);
+        let part_writer = FileWriter::<ManifestDescribing>::try_new(
```
[the remark this answers is not in the document]
Comment (BubbleCal) [reply rust/lance/src/index/vector/ivf/io.rs:313]: fixed by storing these files in tmp dir

File: rust/lance/src/index/vector/ivf/io.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance/src/index/vector/ivf/io.rs]: it's up to how many cores and memory we have, something like `min(num_cpus, memory / partition_size)`

**Review by eddyxu — APPROVED** [review]

**Pull request merged**